### DISCUSSION TITLE
Support tool for dashscope streaming model

### DISCUSSION THREAD

**Discussion by BeamLiu** [langchain4j/langchain4j#1994] [Ideas]
hi there, the QwenStreamingChatModel doesn't support tool, is there any plan to support it recently?

**Comment by jiangsier-xyz** [comment]
> hi there, the QwenStreamingChatModel doesn't support tool, is there any plan to support it recently?

Bro, I also hope to implement this feature. However, according to the [documentation](https://help.aliyun.com/zh/dashscope/developer-reference/qwen-api#ea3869ca53mh4), the model does not support it yet.

Note that the documentation includes the following description: "tools暂时无法与stream=True同时使用" (parameter tools cannot be used with stream=True at this time)

**Reply by BeamLiu** [reply]
Thanks, but it seems the Qwen itself supports tool when streaming. To simplify it, I had some codesnippet with python, it does WORK. But we need carefully handle the tool_calls part message,  as it is also streaming. merge them and finally invoke function.

https://github.com/BeamLiu/codesnippet/blob/main/dashscope_stream_tool.py

At last,  it seems a bit tricky, I don't know how the OpenAI streaming call to handle the tool in stream chat. This is an important feature to me. In a chat environment, frontend cannot know whether the tool function call is necessary, otherwise, I can only choose to use none-streaming, there will be a big latence in frontend. Or there is any other good practise?

**Reply by jiangsier-xyz** [reply]
> Thanks, but it seems the Qwen itself supports tool when streaming. To simplify it, I had some codesnippet with python, it does WORK. But we need carefully handle the tool_calls part message, as it is also streaming. merge them and finally invoke function.
> 
> https://github.com/BeamLiu/codesnippet/blob/main/dashscope_stream_tool.py
> 
> At last, it seems a bit tricky, I don't know how the OpenAI streaming call to handle the tool in stream chat. This is an important feature to me. In a chat environment, frontend cannot know whether the tool function call is necessary, otherwise, I can only choose to use none-streaming, there will be a big latence in frontend. Or there is any other good practise?

OK. I will try to implement tool-calls in QwenStreamingChatModel. If it can pass the tests, means that the documentation is lagging behind the technology.

**Reply by BeamLiu** [reply]
thanks, looking forward to that.
